### PR TITLE
Updating schedule should recalculate next session time

### PR DESCRIPTION
**Pull request by frsyuki**
Schedule was not recalculated when a new revision is pushed. This
becomes a problem when a previous revision has a daily schedule by
acccident and new revision fixes it to hourly schedule.

### CODE REVIEW

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java [hunk 1]
```diff
@@ -443,51 +446,54 @@ public StoredWorkflowDefinition insertWorkflowDefinition(int projId, int revId,
...
-                        .execute();
-                    if (n <= 0) {
-                        // TODO exception?
+                Integer matchingSchedId = oldScheduleNames.get(schedule.getWorkflowName());
+                if (matchingSchedId != null) {
+                    // found the same name. lock it and update
+                    ScheduleStatus status = dao.lockScheduleById(matchingSchedId);
+                    if (status != null) {
+                        ScheduleTime newSchedule = func.apply(status, schedule);
+                        dao.updateScheduleById(
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java [hunk 2]
```diff
@@ -443,51 +446,54 @@ public StoredWorkflowDefinition insertWorkflowDefinition(int projId, int revId,
...
-                            " set workflow_definition_id = :workflowDefinitionId, updated_at = now()" +
-                            // TODO should here update next_run_time nad next_schedule_time?
-                            " where id = :id"
-                        )
-                        .bind("workflowDefinitionId", schedule.getWorkflowDefinitionId())
-                        .bind("id", oldNames.get(schedule.getWorkflowName()))
-                        .execute();
-                    if (n <= 0) {
-                        // TODO exception?
+                Integer matchingSchedId = oldScheduleNames.get(schedule.getWorkflowName());
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java]: [minor] `matched...` would be better?

File: digdag-core/src/main/java/io/digdag/core/repository/ProjectControl.java [hunk 3]
```diff
@@ -98,22 +100,30 @@ private void updateSchedules(
...
         }
 
         // TODO validate workflows and sessions
         //   * compile workflow
         //   * validate SubtaskMatchPattern
 
-        // DatabaseProjectStoreManager.updateSchedules has more logic
-        // to match names with existent schedules
-        store.updateSchedules(project.getId(), schedules.build());
+        store.updateSchedules(project.getId(), schedules.build(), (oldStatus, newSched) -> {
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/repository/ProjectControl.java]: [minor] Just passing `wfIdToSchedulers` as a parameter is simpler?

File: digdag-core/src/main/java/io/digdag/core/repository/ProjectControlStore.java [hunk 4]
```diff
@@ -16,7 +19,12 @@ void insertRevisionArchiveData(int revId, byte[] data)
     StoredWorkflowDefinition insertWorkflowDefinition(int projId, int revId, WorkflowDefinition workflow, ZoneId workflowTimeZone)
         throws ResourceConflictException;
 
-    void updateSchedules(int projId, List<Schedule> schedules)
+    interface ScheduleUpdateAction
+    {
+        public ScheduleTime apply(ScheduleStatus oldStatus, Schedule newSchedule);
```
Comment (komamitsu) [inline digdag-core/src/main/java/io/digdag/core/repository/ProjectControlStore.java]: You can remove `public`

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java [hunk 5]
```diff
@@ -107,7 +114,14 @@ public void testGetAndNotFounds()
...
+                                ),
+                            (oldStatus, newSched) -> {
+                                // when conflicted (wf1), rollback 60 seconds schedule time and 120 seconds run time here
+                                return ScheduleTime.of(
+                                        oldStatus.getNextScheduleTime().getTime().minusSeconds(60),
+                                        oldStatus.getNextScheduleTime().getRunTime().minusSeconds(120));
+                            });
                     return lock.get();
                 });
         StoredRevision rev2 = revRef.get();
```
Comment (komamitsu) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java]: Unused value

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java [hunk 6]
```diff
@@ -79,7 +81,10 @@ public void testGetAndNotFounds()
...
                                     runTime1,
                                     schedTime1)
-                                ));
+                                ),
+                            (oldStatus, newSched) -> {
+                                return oldStatus.getNextScheduleTime();
+                            });
                     return lock.get();
                 });
         StoredRevision rev1 = revRef.get();
```
Comment (komamitsu) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java]: Unused value

**Comment by komamitsu** [comment]
👍 
Left some minor comments, though

File: digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java (hunk 2 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/database/DatabaseProjectStoreManager.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/repository/ProjectControlStore.java (hunk 4 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/repository/ProjectControlStore.java]: 👍

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java (hunk 6 above)
Comment (frsyuki) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java]: 👍

File: digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java (hunk 5 above)
Comment (frsyuki) [inline digdag-core/src/test/java/io/digdag/core/database/DatabaseScheduleStoreManagerTest.java]: 👍

File: digdag-core/src/main/java/io/digdag/core/repository/ProjectControl.java (hunk 3 above)
Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/repository/ProjectControl.java]: I also considered it but I found that the code becomes confusing...`ProjectControlStore.updateSchedules` wants to get list of `Schedule` and callback (ScheduleUpdateAction) wants to use `Scheduler`. To make it possible, here needs a class that extends Schedule and has a Scheduler...

Comment (frsyuki) [inline digdag-core/src/main/java/io/digdag/core/repository/ProjectControl.java]: hm. it was not bad than I expected. So I added 7d3a53a08502ec3a46905602dff67c013b0c06c8.

**Pull request merged**